src/ripple/ledger/CachingReadView.h

```diff
@@ -0,0 +1,149 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_LEDGER_CACHINGREADVIEW_H_INCLUDED
+#define RIPPLE_LEDGER_CACHINGREADVIEW_H_INCLUDED
+
+#include <ripple/ledger/CachedSLEs.h>
+#include <ripple/ledger/ReadView.h>
+#include <ripple/basics/hardened_hash.h>
+#include <map>
+#include <memory>
+#include <mutex>
+
+namespace ripple {
+
+//------------------------------------------------------------------------------
+
+/** ReadView that caches by key and hash. */
+class CachingReadView
+    : public ReadView
+{
+private:
+    CachedSLEs& cache_;
+    std::mutex mutable mutex_;
+    DigestAwareReadView const& base_;
+    std::shared_ptr<void const> hold_;
+    std::unordered_map<key_type,
```

Comment:
Perhaps, but there are sufficiently many user defined types that I have grown suspicious of whether or not those containers do the right thing. Here, you know what you're getting. A standard unordered map with a hardened hash function.